flink-table/flink-table-planner/src/test/java/org/apache/flink/table/api/QueryOperationSqlExecutionTest.java

```diff
@@ -0,0 +1,139 @@
+                GroupWindowAggregateTestPrograms.GROUP_HOP_WINDOW_EVENT_TIME,
+                SortTestPrograms.SORT_LIMIT_DESC,
+                GroupAggregateTestPrograms.GROUP_BY_UDF_WITH_MERGE,
+                JoinTestPrograms.NON_WINDOW_INNER_JOIN);
+    }
+
+    @ParameterizedTest
+    @MethodSource("supportedPrograms")
+    void testSerializedSqlExecution(TableTestProgram program)
+            throws ExecutionException, InterruptedException {
+        final TableEnvironment env = setupEnv(program);
+
+        final TableApiTestStep tableApiStep =
+                (TableApiTestStep)
+                        program.runSteps.stream()
+                                .filter(s -> s instanceof TableApiTestStep)
+                                .findFirst()
+                                .get();
+
+        final TableResult tableResult = tableApiStep.applyAsSql(env);
+        tableResult.await();
+
+        program.getSetupSinkTestSteps()
+                .forEach(
+                        s -> {
+                            assertThat(TestValuesTableFactory.getRawResultsAsStrings(s.name))
+                                    .containsExactlyInAnyOrderElementsOf(s.getExpectedAsStrings());
+                        });
+    }
+
+    private static TableEnvironment setupEnv(TableTestProgram program) {
+        final TableEnvironment env = TableEnvironment.create(EnvironmentSettings.inStreamingMode());
+        final Map<String, String> connectorOptions = new HashMap<>();
+        connectorOptions.put("connector", "values");
+        connectorOptions.put("sink-insert-only", "false");
+        connectorOptions.put("runtime-source", "NewSource");
+        program.getSetupSourceTestSteps()
+                .forEach(
+                        s -> {
+                            final List<Row> data = new ArrayList<>(s.dataBeforeRestore);
+                            data.addAll(s.dataAfterRestore);
+                            final String id = TestValuesTableFactory.registerData(data);
+                            connectorOptions.put("data-id", id);
+                            s.apply(env, connectorOptions);
+                        });
+        program.getSetupSinkTestSteps().forEach(s -> s.apply(env, connectorOptions));
+        program.getSetupFunctionTestSteps().forEach(f -> f.apply(env));
+        return env;
+    }
+
+    @Override
+    public EnumSet<TestKind> supportedSetupSteps() {
+        return EnumSet.of(
+                TestKind.FUNCTION,
+                TestKind.SOURCE_WITH_DATA,
+                TestKind.SOURCE_WITHOUT_DATA,
+                TestKind.SOURCE_WITH_RESTORE_DATA, // restore data is ignored
+                TestKind.SINK_WITH_DATA,
+                TestKind.SINK_WITH_RESTORE_DATA // restore data is ignored
+                );
+    }
+
+    @Override
+    public EnumSet<TestKind> supportedRunSteps() {
+        return EnumSet.of(TestKind.TABLE_API, TestKind.SQL);
```

Comment:
There is no code for supporting TestKind.SQL.

Reply:
Correct, the reason is we don't care about `SQL` here, but if we want to reuse `TestPrograms` that have both `SQL` and `TableAPI` we must declare it in the `supportedRunSteps`.

Reply:
but can we add this when we need it? right it doesn't contain SQL steps, right?

Reply:
It does. `TestPrograms` used here, have the `runSql` which is used for verifying SQL in `QueryOperationSqlSerializationTest`